README.md

```diff
@@ -75,6 +75,73 @@ The PMEM-CSI driver supports running in different modes, which can be controlled
 
 * **_Unified_** mode is intended to run the driver in a single host, mostly for testing the driver in a non-clustered environment.
 
+### Volume Persistency
+
+Normally, volumes are provided by a storage backend that is independent
+of a particular node. When a node goes offline, the volume can be
+mounted elsewhere. With PMEM-CSI, volumes are *local* and thus can
+only be used on the node where they were created.
+
+This limitation needs to be considered when designing and deploying
+applications that are to use local storage:
+
+* Persistent Volumes  
+A volume gets created independently of the application, on some node
+where there is enough free space. Applications using such a volume are
+then forced to run on that node and cannot run when the node is
+down. Data is retained until the volume gets deleted.
+* Ephemeral Volumes  
```

Comment:
@pohly What is the expectation from driver to support this volume type.

As of my understanding there is no way driver knows when the pod existed. To support this usecase CO must initiate DevelVolume request then driver aromatically delete the volume . Unless external component call this upon Pod exit, driver itself can do nothing.

Reply:
One possible implementation of an ephemeral volume is this:
- `CreateVolume`: return a topology that maps to all nodes where PMEM is available (in the sense of "it might be available when a pod runs", not as in "available right now")
- `NodeStageVolume`: create the volume, format and mount it
- `NodePublishVolume`: bind-mount into pod
- `NodeUnpublishVolume`: remove bind-mound
- `NodeUnstageVolume`: delete the volume

I am aware that this may lead to scheduling pods onto a node where currently no local storage is available. There is no good solution for that, we need to work with upstream to enhance the Kubernetes scheduler or write our own.

Reply:
What if we, at CreateVolume, using pmem-csi internal comms, actually reserve the space on those nodes having enough free space? Then the later-following StageVolume can't fail on any of those nodes as there is namespace waiting. After one node starts using that Volume, controller will need to order freeing reservations on other nodes. That would help with "oops there was no space when we really started to create one". Downside is, nodes will perform extra allocations and have to deal with removing those, with added complexity with more error-paths. And we should add some garbage-orphan-checking pass (at plugin startup?) to detect and clean unused left-over namespaces, which is actually good idea to have anyway (we dont have such now).

Reply:
The conceptual model is that there is one PVC that describes a volume of a certain size. Then each pod that uses that PVC gets its own local instance of that size.

That doesn't work when we reserve space once for one pod when creating the PVC and then free the reservation once that pod runs.

Reply:
As of my knowledge "ephemeral" volumes concept is for Pods requesting "in-line" volume claims( i know its not yet supported for csi volumes), and  are owed by single Pod, and cannot be shared by two applications, that means sharing of PVC/PV  concept does not apply here.

Ephemeral requests should reach us only after once Scheduler choose the node for the requested Pod, and we create the volume on that node, so no need of guess work here by driver.
And we strictly *deny* reattachment requests for these volumes and clean the volume on `ControllerUnpublish` unlike normal volumes where we clear on `DeleteVolume`.

If i want to say in Kuberntes terms we only support "ephemeral" volumes that are created by `WaitForFirstConsumer` StorageClass

@pohly adn @okartau can we agree to this?

Reply:
@avalluri yes, ideally "ephemeral volumes" should be implemented as inline volumes. But inline volumes are in design at the moment upstream and thus not available to us. What I described and what @okartau commented on is my proposal how we can approximate the concept with a fake PV. `WaitForFirstConsumer` doesn't help, because we will have multiple consumers of that PV.

As we agreed yesterday to focus on "cache volumes", we don't need to think too hard about how to implement the other persistency models - at least not until it becomes urgent.